File: src/Core/FinancialEvents/Schemas/Lots.py
from dataclasses import dataclass, field
from typing import Generic, Sequence, TypeVar

from Core.FinancialEvents.Schemas.CommonFormats import GenericShortLong
from Core.FinancialEvents.Schemas.Events import (
    TradeEvent,
    TradeEventDerivativeAcquired,
    TradeEventDerivativeSold,
    TradeEventStockAcquired,
    TradeEventStockSold,
)
from Core.FinancialEvents.Schemas.FinancialIdentifier import FinancialIdentifier
from Core.FinancialEvents.Schemas.Provenance import ProvenanceStep

TaxLotAcquiredEvent = TypeVar("TaxLotAcquiredEvent", bound=TradeEvent, covariant=True)
TaxLotSoldEvent = TypeVar("TaxLotSoldEvent", bound=TradeEvent, covariant=True)


@dataclass
class TaxLot(Generic[TaxLotAcquiredEvent, TaxLotSoldEvent]):
    ID: str
    FinancialIdentifier: FinancialIdentifier
    Quantity: float
    Acquired: TaxLotAcquiredEvent
    Sold: TaxLotSoldEvent
    ShortLongType: GenericShortLong  # Some trades can be SHORTING, meaning you first sell and then buy back
    Provenance: Sequence[ProvenanceStep] = field(default_factory=list)


@dataclass
class TaxLotStock(TaxLot[TradeEventStockAcquired, TradeEventStockSold]): ...


@dataclass
class TaxLotDerivative(TaxLot[TradeEventDerivativeAcquired, TradeEventDerivativeSold]): ...

File: lib/core/schemas/Lots.ts
import type { GenericShortLong } from "./CommonFormats.ts";
import type { FinancialIdentifier } from "./FinancialIdentifier.ts";
import type { AnyProvenanceStep } from "./Provenance.ts";
import type { TradeEventDerivativeAcquired, TradeEventDerivativeSold, TradeEventStockAcquired, TradeEventStockSold } from "./Events.ts";

export type TaxLot<A, S> = {
	id: string;
	financialIdentifier: FinancialIdentifier;
	quantity: number;
	acquired: A;
	sold: S;
	shortLongType: GenericShortLong;
	provenance: AnyProvenanceStep[];
};

export type TaxLotStock = TaxLot<TradeEventStockAcquired, TradeEventStockSold>;

export type TaxLotDerivative = TaxLot<TradeEventDerivativeAcquired, TradeEventDerivativeSold>;
